xui/additional_files_unzipped/s3_browser/s3_browser/aws_bucket_browser_operations/aws_bucket_browser_operations.py

```diff
@@ -0,0 +1,343 @@
+            )
+
+            item_type = "Folder"  # for show template
+            url = folder_prefix
+            folder_obj = {
+                            "key": folder_prefix,
+                            "url": url,
+                            "name": folder_prefix.replace(main_folder, ""),
+                            "item_type": item_type
+                        }
+            if is_folder_deleted:
+                folder_obj['is_delete_marker'] = True
+            files_list.append(folder_obj)
+
+
+        return sorted(
+            files_list, key=lambda k: str(k["key"]).lower(), reverse=not sort_a_z
+        )
+    except Exception as e:
+        print(
+            "Error on line {}".format(
+                sys.exc_info()[-1].tb_lineno), type(e).__name__, e
+        )
+
+
+def is_versioning_enabled(s3_client, resource_id, default_region):
+    """
+        Check if versioning is enabled for an S3 bucket linked to a specified resource.
+
+        This function attempts to fetch a resource and its associated AWSHandler based on the provided 
+        `resource_id`. It then tries to access a custom field named "s3_bucket_region" and use its value 
+        to create an S3 resource client. Finally, it checks whether versioning is enabled for the S3 bucket 
+        associated with the resource.
+
+        Parameters:
+        - s3_client (boto3.client): A boto3 S3 client.
+        - resource_id (int or str): The ID of the resource which is linked to the S3 bucket.
+        - default_region: default s3 region passed from the IWH S3 Browser
+
+        Returns:
+        bool: True if versioning is enabled for the S3 bucket, otherwise False.
+
+        Note: 
+        The function returns False if any error/exception occurs during the process.
+    """
+    
+    try:
+        # Get the resource by ID
+        resource = get_object_or_404(Resource, pk=resource_id)
+        aws = get_object_or_404(AWSHandler, pk=resource.aws_rh_id)
+
+        # Get the region from the resource
+        s3rcf = CustomField.objects.get(name="s3_bucket_region")
+
+        region_attribute = resource.attributes.get(field_id=s3rcf.id)
+        region = region_attribute.str_value if region_attribute and region_attribute.str_value else default_region # default_region set from bucket_location in IWH 'S3 Browser'
+        # Get a boto3 resource wrapper for "bucket"
+        s3_bucket = aws.get_boto3_resource(region_name=region, service_name="s3")
+
+        if s3_bucket.BucketVersioning(resource.name).status == "Enabled":
+            return True
+        return False
+
+    except (CustomField.DoesNotExist, CustomFieldValue.DoesNotExist) as error:
+        # This error can arise when s3 buckets are added from discovery. In which case, attempt to find versioning a last with using the default region
+        logger.error(f"CustomField 's3_bucket_region' does not exist. Trying default_region={default_region}")
```

Comment:
should this really be an error, or just a .warning() or .info()?

Reply:
i think .warning() since its something that you can fix